src/charm.py

```diff
@@ -89,30 +229,52 @@ def _on_postgresql_pebble_ready(self, event: WorkloadEvent) -> None:
         # Create a new config layer.
         new_layer = self._postgresql_layer()
 
-        if container.can_connect():
-            # Get the current layer.
-            current_layer = container.get_plan()
-            # Check if there are any changes to layer services.
-            if current_layer.services != new_layer.services:
-                # Changes were made, add the new layer.
-                container.add_layer(self._postgresql_service, new_layer, combine=True)
-                logging.info("Added updated layer 'postgresql' to Pebble plan")
-                # TODO: move this file generation to on config changed hook
-                # when adding configs to this charm.
-                self._patroni.render_patroni_yml_file()
-                # Restart it and report a new status to Juju.
-                container.restart(self._postgresql_service)
-                logging.info("Restarted postgresql service")
-            # All is well, set an ActiveStatus.
-            self.unit.status = ActiveStatus()
-        else:
+        # Defer the event if we pebble is not available yet.
+        if not container.can_connect():
             self.unit.status = WaitingStatus("waiting for Pebble in workload container")
+            event.defer()
+            return
+
+        # Defer the initialization of the workload in the replicas
+        # if the cluster hasn't been bootstrap on the primary yet.
+        if not self.unit.is_leader() and "cluster_initialised" not in self._peers.data[self.app]:
+            event.defer()
+            return
```

Comment:
+1 here. If the member are added on the `relation_changed`, why delay initialisation?

Reply:
I added a better comment about that on https://github.com/canonical/postgresql-k8s-operator/pull/9/commits/ee0018899c32c6f9b60debcc5179ef39eb6ac02c. It's needed because if multiple units start Patroni at the same time they will create multiple clusters (separated from each other). We need to start the primary first and link the other units to it (through the cluster members list) before starting.